src/sentry/snuba/discover.py

```diff
@@ -0,0 +1,230 @@
+    raw_query,
+)
+
+ALLOWED_INTERNALS = set(["project_id", "event_id"])
+
+
+def resolve_column(col):
+    """
+    Resolve a public schema name to the discover dataset.
+    unknown columns are converted into tags expressions.
+    """
+    if not col:
+        return ""
+    if col.startswith("tags[") or col in ALLOWED_INTERNALS or QUOTED_LITERAL_RE.match(col):
+        return col
+    return DISCOVER_COLUMN_MAP.get(col, u"tags[{}]".format(col))
+
+
+def resolve_condition(cond):
+    """
+    When conditions have been parsed by the api.event_search module
+    we can end up with conditions that are not valid on the current dataset
+    due to how ap.event_search checks for valid field names without
+    being aware of the dataset.
+
+    We have the dataset context here, so we need to re-scope conditions to the
+    current dataset.
+    """
+    index = get_function_index(cond)
+    if index is not None:
+        # IN conditions are detected as a function but aren't really.
+        if cond[index] == "IN":
+            cond[0] = resolve_column(cond[0])
+            return cond
+
+        func_args = cond[index + 1]
+        for (i, arg) in enumerate(func_args):
+            # Nested function
+            if isinstance(arg, (list, tuple)):
+                func_args[i] = resolve_condition(arg)
+            else:
+                func_args[i] = resolve_column(arg)
+        cond[index + 1] = func_args
+        return cond
+
+    # No function name found
+    if isinstance(cond, (list, tuple)) and len(cond):
+        # Condition is [col, operator, value]
+        if isinstance(cond[0], six.string_types) and len(cond) == 3:
+            cond[0] = resolve_column(cond[0])
+            return cond
+        if isinstance(cond[0], (list, tuple)):
+            if get_function_index(cond[0]) is not None:
+                cond[0] = resolve_condition(cond[0])
+                return cond
+            else:
+                # Nested conditions
+                return [resolve_condition(item) for item in cond]
+    raise ValueError("Unexpected condition format %s" % cond)
+
+
+def resolve_discover_aliases(snuba_args):
+    """
+    Resolve the public schema aliases to the discover dataset.
+
+    Returns a copy of the input structure, and includes a
+    `translated_columns` key containing the selected fields that need to
+    be renamed in the result set.
+    """
+    resolved = deepcopy(snuba_args)
+    translated_columns = {}
+    derived_columns = set()
+
+    selected_columns = resolved.get("selected_columns")
+    if selected_columns:
+        for (idx, col) in enumerate(selected_columns):
+            if isinstance(col, (list, tuple)):
+                # if list, means there are potentially nested functions and need to
+                # iterate and translate potential columns which is the same process
+                # as handling a condition.
+                selected_columns[idx] = resolve_condition(col, Dataset.Discover)
```

Comment:
@markstory `Dataset.Discover` is passed as a 2nd argument to `resolve_condition()`. Is this a typo?

Reply:
It is, I'll get that fixed.